CMakeLists.txt

```diff
@@ -1,431 +1,507 @@
 endif()
 
-include_directories(${PROJECT_SOURCE_DIR})
-find_package(Threads REQUIRED)
-message(STATUS "Found Threads: ${CMAKE_THREAD_LIBS_INIT}")
-
-# Include Submodules
-add_subdirectory(external_tools/googletest EXCLUDE_FROM_ALL)
-include_directories(SYSTEM ${gtest_SOURCE_DIR}/include ${gtest_SOURCE_DIR})
-include_directories(SYSTEM ${gtest_SOURCE_DIR}/../googlemock/include ${gtest_SOURCE_DIR}/../googlemock/)
-include_directories(external_tools/kahypar-shared-resources)
-include_directories(external_tools/growt)
-include_directories(external_tools/WHFC)
-include_directories(external_tools/pcg)
-
-if(KAHYPAR_DOWNLOAD_BOOST)
-  # Download Boost
-  execute_process(COMMAND cmake -P ${CMAKE_CURRENT_SOURCE_DIR}/scripts/download_boost.cmake)
-  include_directories(SYSTEM ${CMAKE_CURRENT_BINARY_DIR}/external_tools/boost/)
-  file(GLOB MINI_BOOST_SOURCES ${CMAKE_CURRENT_BINARY_DIR}/external_tools/boost/libs/program_options/src/*.cpp)
-
-  add_library(mini_boost STATIC ${MINI_BOOST_SOURCES})
-  set_target_properties(mini_boost PROPERTIES LINKER_LANGUAGE CXX)
-  set(Boost_LIBRARIES mini_boost)
-  set(Boost_INCLUDE_DIRS ${CMAKE_CURRENT_BINARY_DIR}/external_tools/boost/boost/)
-  include_directories(SYSTEM ${Boost_INCLUDE_DIRS})
-else()
-  # Find Boost Program Options Library
-  find_package(Boost 1.69 REQUIRED COMPONENTS program_options)
-  if(Boost_FOUND)
-    include_directories(SYSTEM ${Boost_INCLUDE_DIRS})
-    set(KAHYPAR_INCLUDE_DIRS ${Boost_INCLUDE_DIRS} ${KAHYPAR_INCLUDE_DIRS})
-    set(KAHYPAR_LINK_LIBRARIES ${Boost_LIBRARIES} ${KAHYPAR_LINK_LIBRARIES})
-    message(STATUS "Boost Include: ${Boost_INCLUDE_DIRS}")
-    message(STATUS "Boost Library Dirs: ${Boost_LIBRARY_DIRS}")
-    message(STATUS "Boost Libraries: ${Boost_LIBRARIES}")
-    if(WIN32)
-      add_definitions(-DBOOST_ALL_NO_LIB)
-      add_definitions(-DBOOST_PROGRAM_OPTIONS_DYN_LINK=1)
-    endif()
+
+#################################################################
+## Setup of dependencies                                       ##
+#################################################################
+
+include(FetchContent)
+set(KAHYPAR_TBB_VERSION        v2022.0.0)
+set(KAHYPAR_GOOGLETEST_VERSION v1.15.2)
+set(KAHYPAR_PYBIND11_VERSION   v2.13.6)
+
+message(STATUS "Fetching dependencies...")
+
+# Include header-only dependencies
+FetchContent_Populate(
+  kahypar-shared-resources QUIET EXLUDE_FROM_ALL
+  GIT_REPOSITORY https://github.com/kahypar/kahypar-shared-resources.git
+  GIT_TAG        origin/main
+  SOURCE_DIR     external_tools/kahypar-shared-resources
+)
+FetchContent_Populate(
+  WHFC QUIET EXLUDE_FROM_ALL
+  GIT_REPOSITORY https://github.com/larsgottesbueren/WHFC.git
+  GIT_TAG        origin/master
+  SOURCE_DIR     external_tools/WHFC
+)
+FetchContent_Populate(
+  growt QUIET EXLUDE_FROM_ALL
+  GIT_REPOSITORY https://github.com/TooBiased/growt.git
+  GIT_TAG        origin/master
+  SOURCE_DIR     external_tools/growt
+)
```

Comment:
This always checks out latest `master` from the repositories which is kind of dangerous if there are new commits that could potentially break some code on our side. Better checkout specific git commit.

Reply:
All of these dependencies are quite stable and unlikely to make any breaking changes. `kahypar-shared-resources` and `whfc` are even controlled by us, so for these it should not be an issue at all. In practice, it seems much more likely to me that someone forgets to update the git commit here after submitting a bug fix for one of the dependecies, which is avoided by checking out `master`

Reply:
Quite stable and unlikely to break anything means that there is still a small probability that these repository can change and break our code. I understand your point. However, checking out master per default means that someone can push something to an external dependency without any validation by our internal CI. Also bug fixes in external dependencies needs to be first validated by our CI. Therefore, migrating to a new version of an external dependency should be done explicitly.